distributed/worker.py

```diff
@@ -266,32 +270,48 @@ def _start(self, addr_or_port=0):
         # XXX Factor this out
         if not addr_or_port:
             # Default address is the required one to reach the scheduler
-            self.listen(get_local_address_for(self.scheduler.address),
+            listen_host = get_local_address_for(self.scheduler.address)
+            self.listen(listen_host,
                         listen_args=self.listen_args)
             self.ip = get_address_host(self.address)
         elif isinstance(addr_or_port, int):
             # addr_or_port is an integer => assume TCP
             self.ip = get_ip(
                 get_address_host(self.scheduler.address)
             )
-            self.listen((self.ip, addr_or_port),
+            self.listen((listen_host, addr_or_port),
                         listen_args=self.listen_args)
+            listen_host = self.ip
         else:
             self.listen(addr_or_port, listen_args=self.listen_args)
             self.ip = get_address_host(self.address)
+            try:
+                listen_host = get_address_host(addr_or_port)
+            except ValueError:
+                listen_host = addr_or_port
+
+        if '://' in listen_host:
+            protocol, listen_host = listen_host.split('://')
+            if protocol == 'inproc':
+                listen_host = 'localhost'
```

Comment:
Can you explain why this is required?
Also, at this point, is `self.ip` still used for something?